jupyterhub/tests/selenium/test_browser.py

```diff
@@ -0,0 +1,228 @@
+import asyncio
+import time
+from functools import partial
+
+import pytest
+from selenium import webdriver
+from selenium.common.exceptions import NoSuchElementException, TimeoutException
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.ui import WebDriverWait
+from tornado.escape import url_escape
+from tornado.httputil import url_concat
+
+from jupyterhub.tests.selenium.locators import (
+    HomePageLocators,
+    LoginPageLocators,
+    TokenPageLocators,
+)
+from jupyterhub.utils import exponential_backoff
+
+from ...utils import url_path_join
+from ..utils import public_host, public_url, ujoin
+
+pytestmark = pytest.mark.selenium
+
+
+async def webdriver_wait(driver, condition, timeout=30):
+    return await exponential_backoff(
+        partial(condition, driver),
+        timeout=timeout,
+        fail_message=f"WebDriver condition not met: {condition}",
+    )
+
+
+def in_thread(f, *args, **kwargs):
+    """Run a function in a background thread
+
+    via current event loop's run_in_executor
+
+    Returns asyncio.Future
+    """
+
+    return asyncio.get_event_loop().run_in_executor(None, partial(f, *args, **kwargs))
+
+
+async def open_url(app, browser):
+    """initiating open the login page in the browser"""
+    url = url_path_join(public_host(app), app.hub.base_url, "login")
+    await in_thread(browser.get, url)
+    return url
+
+
+def click(browser, by_locator):
+    WebDriverWait(browser, 10).until(
```

Comment:
Add a docstring here and below about "wait for element to be visible, then click on it"